src/iron.rs

```diff
@@ -111,89 +128,139 @@ impl<H: Handler> Iron<H> {
     /// Kick off the server process using the HTTP protocol.
     ///
     /// Call this once to begin listening for requests on the server.
-    /// This consumes the Iron instance, but does the listening on
-    /// another task, so is not blocking.
-    ///
-    /// The thread returns a guard that will automatically join with the parent
-    /// once it is dropped, blocking until this happens.
-    pub fn http<A>(self, addr: A) -> HttpResult<Listening>
+    pub fn http<A>(self, addr: A) -> HttpResult<()>
         where A: ToSocketAddrs
     {
-        HttpListener::new(addr).and_then(|l| self.listen(l, Protocol::http()))
+        let addr = addr.to_socket_addrs()?.next().unwrap();
+
+        let core = Core::new().unwrap();
+        let handle = core.handle();
+
+        let sock = TcpListener::bind(&addr, &handle).unwrap().incoming();
+
+        return self.listen(sock, addr, Protocol::http(), core, handle);
     }
 
     /// Kick off the server process using the HTTPS protocol.
     ///
     /// Call this once to begin listening for requests on the server.
-    /// This consumes the Iron instance, but does the listening on
-    /// another task, so is not blocking.
-    ///
-    /// The thread returns a guard that will automatically join with the parent
-    /// once it is dropped, blocking until this happens.
-    pub fn https<A, S>(self, addr: A, ssl: S) -> HttpResult<Listening>
-        where A: ToSocketAddrs,
-              S: 'static + SslServer + Send + Clone
+    #[cfg(feature = "ssl")]
+    pub fn https<A, Tls>(self, addr: A, tls: Tls) -> HttpResult<()>
+        where A: ToSocketAddrs, Tls: TlsAcceptorExt + 'static
     {
-        HttpsListener::new(addr, ssl).and_then(|l| self.listen(l, Protocol::http()))
+        let addr = addr.to_socket_addrs()?.next().unwrap();
+
+        let core = Core::new().unwrap();
+        let handle = core.handle();
+
+        let listener_tcp = TcpListener::bind(&addr, &handle).unwrap();
+
+        let (tx, rx) = mpsc::channel(1);
+
+        let ssl_acceptor = listener_tcp.incoming().for_each(move |(sock, remote_addr)| {
+            let tx = tx.clone();
+            tls.accept_async(sock).map_err(|e| IoError::new(ErrorKind::Other, e)).and_then(move |sock| {
+                future::ok((sock, remote_addr))
+            }).then(|r| {
+                tx.send(r).map_err(|e| IoError::new(ErrorKind::Other, e))
+            }).and_then(|_| future::ok(()))
+        }).then(|_| future::ok(()));
+        handle.spawn(ssl_acceptor);
+
+        let listener = rx.then(|r| match r {
+            Ok(real_r) => real_r,
+            Err(e) => panic!(e),
```

Comment:
I believe this operation can "never" fail, so I included a panic to signify this.  But I can't find any documentation specifically saying this, just that the error type of r is the unit type.  I'll dig into this more sometime this week to verify.